File: sdks/ios/Sources/RRA/RRAClient.swift
// SPDX-License-Identifier: FSL-1.1-ALv2
// Copyright 2025 Kase Branham

import Foundation

/// Main client for interacting with the RRA (Revenant Repo Agent) API.
///
/// Example usage:
/// ```swift
/// let client = RRAClient(baseURL: "https://api.natlangchain.io", apiKey: "your-api-key")
/// let agents = try await client.marketplace.listAgents()
/// ```
public class RRAClient {

    // MARK: - Properties

    /// Base URL for the RRA API
    public let baseURL: URL

    /// API key for authentication
    private let apiKey: String?

    /// URL session for network requests
    private let session: URLSession

    /// JSON encoder for requests
    private let encoder: JSONEncoder

    /// JSON decoder for responses
    private let decoder: JSONDecoder

    // MARK: - Sub-clients

    /// Marketplace API client
    public lazy var marketplace: MarketplaceClient = {
        MarketplaceClient(client: self)
    }()

    /// Negotiation API client
    public lazy var negotiation: NegotiationClient = {
        NegotiationClient(client: self)
    }()

    /// Streaming payments API client
    public lazy var streaming: StreamingClient = {
        StreamingClient(client: self)
    }()

    /// Analytics API client
    public lazy var analytics: AnalyticsClient = {
        AnalyticsClient(client: self)
    }()

    /// Widget API client
    public lazy var widget: WidgetClient = {
        WidgetClient(client: self)
    }()

    // MARK: - Initialization

    /// Initialize the RRA client.
    ///
    /// - Parameters:
    ///   - baseURL: Base URL for the RRA API
    ///   - apiKey: Optional API key for authentication
    ///   - session: URL session to use (defaults to shared)
    public init(
        baseURL: String,
        apiKey: String? = nil,
        session: URLSession = .shared
    ) {
        self.baseURL = URL(string: baseURL)!
        self.apiKey = apiKey
        self.session = session

        self.encoder = JSONEncoder()
        self.encoder.keyEncodingStrategy = .convertToSnakeCase
        self.encoder.dateEncodingStrategy = .iso8601

        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase
        self.decoder.dateDecodingStrategy = .iso8601
    }

    // MARK: - Request Methods

    /// Perform a GET request.
    internal func get<T: Decodable>(
        path: String,
        queryItems: [URLQueryItem]? = nil
    ) async throws -> T {
        let request = try buildRequest(method: "GET", path: path, queryItems: queryItems)
        return try await perform(request)
    }

    /// Perform a POST request with a body.
    internal func post<Body: Encodable, Response: Decodable>(
        path: String,
        body: Body
    ) async throws -> Response {
        var request = try buildRequest(method: "POST", path: path)
        request.httpBody = try encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await perform(request)
    }

    /// Perform a DELETE request.
    internal func delete(path: String) async throws {
        let request = try buildRequest(method: "DELETE", path: path)
        let (_, response) = try await session.data(for: request)
        try validateResponse(response)
    }

    // MARK: - Private Methods

    private func buildRequest(
        method: String,
        path: String,
        queryItems: [URLQueryItem]? = nil
    ) throws -> URLRequest {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: true)!
        components.queryItems = queryItems

        guard let url = components.url else {
            throw RRAError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method

        if let apiKey = apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        }

        request.setValue("RRA-iOS-SDK/1.0.0", forHTTPHeaderField: "User-Agent")

        return request
    }

    private func perform<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)
        try validateResponse(response)

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw RRAError.decodingError(error)
        }
    }

    private func validateResponse(_ response: URLResponse) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw RRAError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200...299:
            return
        case 401:
            throw RRAError.unauthorized
        case 403:
            throw RRAError.forbidden
        case 404:
            throw RRAError.notFound
        case 429:
            throw RRAError.rateLimited
        case 500...599:
            throw RRAError.serverError(httpResponse.statusCode)
        default:
            throw RRAError.httpError(httpResponse.statusCode)
        }
    }
}

// MARK: - Errors

/// Errors that can occur when using the RRA SDK.
public enum RRAError: Error, LocalizedError {
    case invalidURL
    case invalidResponse
    case unauthorized
    case forbidden
    case notFound
    case rateLimited
    case serverError(Int)
    case httpError(Int)
    case decodingError(Error)
    case encodingError(Error)
    case websocketError(String)

    public var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid response from server"
        case .unauthorized:
            return "Authentication required"
        case .forbidden:
            return "Access forbidden"
        case .notFound:
            return "Resource not found"
        case .rateLimited:
            return "Rate limit exceeded"
        case .serverError(let code):
            return "Server error: \(code)"
        case .httpError(let code):
            return "HTTP error: \(code)"
        case .decodingError(let error):
            return "Failed to decode response: \(error.localizedDescription)"
        case .encodingError(let error):
            return "Failed to encode request: \(error.localizedDescription)"
        case .websocketError(let message):
            return "WebSocket error: \(message)"
        }
    }
}

File: sdks/android/src/main/kotlin/io/natlangchain/rra/RRAClient.kt
// SPDX-License-Identifier: FSL-1.1-ALv2
// Copyright 2025 Kase Branham

package io.natlangchain.rra

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Main client for interacting with the RRA (Revenant Repo Agent) API.
 *
 * Example usage:
 * ```kotlin
 * val client = RRAClient(
 *     baseUrl = "https://api.natlangchain.io",
 *     apiKey = "your-api-key"
 * )
 * val agents = client.marketplace.listAgents()
 * ```
 */
class RRAClient(
    private val baseUrl: String,
    private val apiKey: String? = null,
    private val httpClient: OkHttpClient = defaultHttpClient()
) {
    companion object {
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

        private fun defaultHttpClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()
    }

    internal val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
        prettyPrint = false
    }

    /** Marketplace API client */
    val marketplace: MarketplaceClient by lazy { MarketplaceClient(this) }

    /** Negotiation API client */
    val negotiation: NegotiationClient by lazy { NegotiationClient(this) }

    /** Streaming payments API client */
    val streaming: StreamingClient by lazy { StreamingClient(this) }

    /** Analytics API client */
    val analytics: AnalyticsClient by lazy { AnalyticsClient(this) }

    /** Widget API client */
    val widget: WidgetClient by lazy { WidgetClient(this) }

    /**
     * Perform a GET request.
     */
    internal suspend inline fun <reified T> get(
        path: String,
        queryParams: Map<String, String>? = null
    ): T = withContext(Dispatchers.IO) {
        val urlBuilder = StringBuilder("$baseUrl$path")
        queryParams?.let { params ->
            if (params.isNotEmpty()) {
                urlBuilder.append("?")
                urlBuilder.append(params.entries.joinToString("&") { "${it.key}=${it.value}" })
            }
        }

        val request = Request.Builder()
            .url(urlBuilder.toString())
            .get()
            .apply { addHeaders(this) }
            .build()

        executeRequest(request)
    }

    /**
     * Perform a POST request with a body.
     */
    internal suspend inline fun <reified B, reified T> post(
        path: String,
        body: B
    ): T = withContext(Dispatchers.IO) {
        val jsonBody = json.encodeToString(body)
        val requestBody = jsonBody.toRequestBody(JSON_MEDIA_TYPE)

        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(requestBody)
            .apply { addHeaders(this) }
            .build()

        executeRequest(request)
    }

    /**
     * Perform a DELETE request.
     */
    internal suspend fun delete(path: String): Unit = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .delete()
            .apply { addHeaders(this) }
            .build()

        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw handleError(response.code, response.body?.string())
            }
        }
    }

    private fun addHeaders(builder: Request.Builder) {
        builder.addHeader("User-Agent", "RRA-Android-SDK/1.0.0")
        apiKey?.let { builder.addHeader("X-API-Key", it) }
    }

    private inline fun <reified T> executeRequest(request: Request): T {
        httpClient.newCall(request).execute().use { response ->
            val body = response.body?.string() ?: ""

            if (!response.isSuccessful) {
                throw handleError(response.code, body)
            }

            return try {
                json.decodeFromString(body)
            } catch (e: Exception) {
                throw RRAException.DecodingError(e)
            }
        }
    }

    private fun handleError(code: Int, body: String?): RRAException {
        return when (code) {
            401 -> RRAException.Unauthorized
            403 -> RRAException.Forbidden
            404 -> RRAException.NotFound
            429 -> RRAException.RateLimited
            in 500..599 -> RRAException.ServerError(code)
            else -> RRAException.HttpError(code, body)
        }
    }
}

/**
 * Exceptions that can occur when using the RRA SDK.
 */
sealed class RRAException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    object Unauthorized : RRAException("Authentication required")
    object Forbidden : RRAException("Access forbidden")
    object NotFound : RRAException("Resource not found")
    object RateLimited : RRAException("Rate limit exceeded")
    class ServerError(code: Int) : RRAException("Server error: $code")
    class HttpError(code: Int, body: String?) : RRAException("HTTP error $code: $body")
    class DecodingError(cause: Throwable) : RRAException("Failed to decode response", cause)
    class EncodingError(cause: Throwable) : RRAException("Failed to encode request", cause)
    class WebSocketError(message: String) : RRAException("WebSocket error: $message")
}
